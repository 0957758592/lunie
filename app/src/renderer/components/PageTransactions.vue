<template>
  <div class="page-transactions">
    <page-header title="Transactions"></page-header>
    <div class="page-content">
      <card-transaction
        v-for="transaction in transactions"
        :transaction-value="transaction">
      </card-transaction>
      <card-empty
        v-if="transactions.length < 1"
        value="No transactions yet">
      </card-empty>
    </div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex'
import { orderBy } from 'lodash'
import PageHeader from './PageHeader'
import CardTransaction from './CardTransaction'
import CardEmpty from './CardEmpty'
export default {
  components: {
    PageHeader,
    CardTransaction,
    CardEmpty
  },
  computed: {
    transactions () {
      return orderBy(this.allTransactions, ['time'], ['desc'])
    },
    ...mapGetters(['allTransactions'])
  }
}
</script>

<style lang="stylus">
@require '../styles/variables.styl'

.page-content
  padding 0.5rem
</style>
