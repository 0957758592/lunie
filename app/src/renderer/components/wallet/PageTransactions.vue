<template lang="pug">
page(title='Transactions')
  div(slot="menu"): tool-bar
    a(@click='setSearch(true)')
      i.material-icons search
      .label Search

  modal-search(v-if="filters.transactions.search.visible" type="transactions")

  data-empty-tx(v-if='transactions.length === 0')
  data-empty-search(v-else-if="filteredTransactions.length")
  li-transaction(
    v-else
    v-for="i in filteredTransactions"
    :transaction-value="i"
    :address="wallet.key.address"
    :devMode="config.devMode")
</template>

<script>
import { mapGetters } from 'vuex'
import { includes, orderBy } from 'lodash'
import Mousetrap from 'mousetrap'
import DataEmptySearch from 'common/NiDataEmptySearch'
import DataEmptyTx from 'common/NiDataEmptyTx'
import LiTransaction from 'wallet/LiTransaction'
import ModalSearch from 'common/NiModalSearch'
import Page from 'common/NiPage'
import Part from 'common/NiPart'
import ToolBar from 'common/NiToolBar'
export default {
  name: 'page-transactions',
  components: {
    LiTransaction,
    DataEmptySearch,
    DataEmptyTx,
    ModalSearch,
    Page,
    Part,
    ToolBar
  },
  computed: {
    ...mapGetters(['filters', 'transactions', 'wallet', 'config']),
    orderedTransactions () {
      return orderBy(this.transactions, [this.sort.property], [this.sort.order])
    },
    filteredTransactions () {
      let query = this.filters.transactions.search.query
      if (this.filters.transactions.search.visible) {
        // doing a full text comparison on the transaction data
        return this.orderedTransactions.filter(t => includes(JSON.stringify(t).toLowerCase(), query))
      } else {
        return this.orderedTransactions
      }
    }
  },
  data: () => ({
    sort: {
      property: 'time',
      order: 'desc'
    }
  }),
  methods: {
    setSearch (bool) {
      this.$store.commit('setSearchVisible', ['transactions', bool])
    }
  },
  mounted () {
    Mousetrap.bind(['command+f', 'ctrl+f'], () => this.setSearch(true))
    Mousetrap.bind('esc', () => this.setSearch(false))
  }
}
</script>
