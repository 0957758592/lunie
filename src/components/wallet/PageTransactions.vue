<template>
  <TmPage
    :managed="true"
    :loading="$apollo.queries.transactions.loading"
    :loaded="!$apollo.queries.transactions.loading"
    :error="$apollo.queries.transactions.error"
    :data-empty="transactions.length === 0"
    data-title="Transactions"
    :sign-in-required="true"
    :hide-header="true"
  >
    <DataEmptyTx slot="no-data" />
    <template slot="managed-body">
      <div infinite-scroll-distance="80">
        <TransactionList
          :transactions="transactions"
          :address="session.address"
          :validators="validatorsAddressMap"
        />
      </div>
      <br />
    </template>
  </TmPage>
</template>

<script>
import { mapState } from "vuex"
import DataEmptyTx from "common/TmDataEmptyTx"
import TmPage from "common/TmPage"
import TransactionList from "transactions/TransactionList"
import gql from "graphql-tag"

export default {
  name: `page-transactions`,
  components: {
    TransactionList,
    DataEmptyTx,
    TmPage
  },
  data: () => ({
    showing: 15,
    validators: [],
    transactions: []
  }),
  computed: {
    ...mapState([`session`]),
    ...mapState({ network: state => state.connection.network }),
    validatorsAddressMap() {
      const names = {}
      this.validators.forEach(item => {
        names[item.operatorAddress] = item
      })
      return names
    },
    dataEmpty() {
      return this.transactions.length === 0
    }
  },
  methods: {
    loadMore() {
      this.showing += 10
    }
  },
  apollo: {
    transactions: {
      query: gql`
        query transactions($networkId: String!, $address: String!) {
          transactions(networkId: $networkId, address: $address) {
            hash
            type
            group
            height
            timestamp
            gasUsed
            fee {
              amount
              denom
            }
            value
          }
        }
      `,
      skip() {
        return !this.session.address
      },
      variables() {
        return {
          networkId: this.network,
          address: this.session.address
        }
      },
      update: result => {
        if (Array.isArray(result.transactions)) {
          return result.transactions.map(transaction => {
            !transaction.value && console.log("TX ERROR")
            const newTx = {
              ...transaction,
              value: JSON.parse(transaction.value),
              timestamp: new Date(transaction.timestamp)
            }
            return newTx
          })
        }
        return []
      }
    },
    validators: {
      query: gql`
        query validators($networkId: String!) {
          validators(networkId: $networkId) {
            name
            operatorAddress
          }
        }
      `,
      variables() {
        return {
          networkId: this.network
        }
      }
    }
  }
}
</script>
