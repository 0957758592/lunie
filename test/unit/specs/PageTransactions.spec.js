import Vuex from 'vuex'
import { mount, createLocalVue } from 'vue-test-utils'
import PageTransactions from 'renderer/components/wallet/PageTransactions'

const wallet = require('renderer/vuex/modules/wallet').default({})
const filters = require('renderer/vuex/modules/filters').default({})

const localVue = createLocalVue()
localVue.use(Vuex)

describe('PageTransactions', () => {
  let wrapper, store

  beforeEach(() => {
    store = new Vuex.Store({
      getters: {
        wallet: () => wallet.state,
        transactions: () => wallet.state.history,
        filters: () => filters.state
      },
      modules: {
        wallet,
        filters
      }
    })

    store.commit('setWalletHistory', [{
      tx: {
        hash: 'x',
        inputs: [
          {
            coins: [{
              denom: 'jbcoins',
              amount: 1234
            }],
            sender: 'myAddress'
          }
        ],
        outputs: [
          {
            coins: [{
              denom: 'jbcoins',
              amount: 1234
            }],
            recipient: 'otherAddress'
          }
        ]
      },
      time: Date.now()
    }, {
      tx: {
        hash: 'y',
        inputs: [
          {
            coins: [{
              denom: 'fabocoins',
              amount: 1234
            }],
            sender: 'otherAddress'
          }
        ],
        outputs: [
          {
            coins: [{
              denom: 'fabocoins',
              amount: 1234
            }],
            recipient: 'myAddress'
          }
        ]
      },
      time: Date.now() + 10
    }])

    wrapper = mount(PageTransactions, {
      localVue,
      store,
      stubs: {
        'data-empty-tx': '<data-empty-tx />'
      }
    })

    jest.spyOn(store, 'commit')
  })

  it('has the expected html structure', () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it('should show the search on click', () => {
    wrapper.find('.ni-tool-bar i').trigger('click')
    expect(wrapper.contains('.ni-modal-search')).toBe(true)
  })

  it('should sort the transaction by time', () => {
    expect(wrapper.vm.filteredTransactions.map(x => x.tx.hash)).toEqual(['y', 'x'])
  })

  it('should filter the transactions', () => {
    store.commit('setSearchVisible', ['transactions', true])
    store.commit('setSearchQuery', ['transactions', 'fabo'])
    expect(wrapper.vm.filteredTransactions.map(x => x.tx.hash)).toEqual(['y'])
    // reflects the filter in the view
    expect(wrapper.vm.$el).toMatchSnapshot()
    store.commit('setSearchQuery', ['transactions', 'jb'])
    expect(wrapper.vm.filteredTransactions.map(x => x.tx.hash)).toEqual(['x'])
  })

  it('should show an error if there are no transactions', () => {
    store.commit('setWalletHistory', [])
    wrapper.update()
    expect(wrapper.contains('data-empty-tx')).toBe(true)
  })
})
