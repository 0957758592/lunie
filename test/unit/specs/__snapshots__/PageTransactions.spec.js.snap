// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTransactions has the expected html structure 1`] = `
<div
  class="ni-page"
>
  <header
    class="ni-page-header"
  >
    <div
      class="ni-page-header-container"
    >
      <div
        class="ni-page-header-text"
      >
        <div
          class="ni-page-header-title"
        >
          <div>
            Transactions
          </div>
        </div>
        <div
          class="ni-page-header-subtitle"
        >
          <div>
            
          </div>
        </div>
      </div>
      <menu
        class="ni-page-header-menu"
      >
        <div>
          <div>
            <div
              class="ni-tool-bar"
            >
              <div
                class="ni-tool-bar-container"
              >
                <div
                  class="main"
                >
                  <a>
                    <i
                      class="material-icons"
                    >
                      search
                    </i>
                    <div
                      class="label"
                    >
                      Search
                    </div>
                  </a>
                </div>
                <a
                  class="help"
                >
                  <i
                    class="material-icons"
                  >
                    help_outline
                  </i>
                </a>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
  </header>
  <main
    class="ni-page-main"
  >
    <!---->
    <div
      class="ni-data-msg"
    >
      <div
        class="ni-data-msg__icon"
      >
        <i
          class="material-icons"
        >
          search
        </i>
      </div>
      <div
        class="ni-data-msg__text"
      >
        <div
          class="ni-data-msg__title"
        >
          <div>
            No Results
          </div>
        </div>
        <div
          class="ni-data-msg__subtitle"
        >
          <div>
            Your search did not match any available data.
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="ni-page-footer"
  >
    <div
      class="container"
    />
  </footer>
</div>
`;

exports[`PageTransactions should filter the transactions 1`] = `
<div
  class="ni-page"
>
  <header
    class="ni-page-header"
  >
    <div
      class="ni-page-header-container"
    >
      <div
        class="ni-page-header-text"
      >
        <div
          class="ni-page-header-title"
        >
          <div>
            Transactions
          </div>
        </div>
        <div
          class="ni-page-header-subtitle"
        >
          <div>
            
          </div>
        </div>
      </div>
      <menu
        class="ni-page-header-menu"
      >
        <div>
          <div>
            <div
              class="ni-tool-bar"
            >
              <div
                class="ni-tool-bar-container"
              >
                <div
                  class="main"
                >
                  <a>
                    <i
                      class="material-icons"
                    >
                      search
                    </i>
                    <div
                      class="label"
                    >
                      Search
                    </div>
                  </a>
                </div>
                <a
                  class="help"
                >
                  <i
                    class="material-icons"
                  >
                    help_outline
                  </i>
                </a>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
  </header>
  <main
    class="ni-page-main"
  >
    <div
      class="ni-modal-search"
    >
      <div
        class="ni-modal-search-container"
      >
        <input
          class="mousetrap ni-field"
          placeholder="Search..."
          type="text"
        />
        <button
          class="ni-btn"
        >
          <span
            class="ni-btn-container"
          >
            <i
              aria-hidden="true"
              class="ni-btn-icon material-icons"
            >
              close
            </i>
            <!---->
          </span>
        </button>
      </div>
    </div>
    <div
      class="ni-data-msg"
    >
      <div
        class="ni-data-msg__icon"
      >
        <i
          class="material-icons"
        >
          search
        </i>
      </div>
      <div
        class="ni-data-msg__text"
      >
        <div
          class="ni-data-msg__title"
        >
          <div>
            No Results
          </div>
        </div>
        <div
          class="ni-data-msg__subtitle"
        >
          <div>
            Your search did not match any available data.
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="ni-page-footer"
  >
    <div
      class="container"
    />
  </footer>
</div>
`;
